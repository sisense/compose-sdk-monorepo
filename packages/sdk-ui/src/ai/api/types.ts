import { JaqlQueryPayload } from '@sisense/sdk-query-client';

import { ExpandedQueryModel } from '@/analytics-composer';
import { WidgetProps } from '@/props';

export interface ChatContext {
  title: string;
  live: boolean;
}

export interface ChatContextDetails {
  dashboardId?: string;
}
export interface TextMessage {
  role: 'assistant';
  content: string; // stringified json { answer: string }
  type: 'text';
}
export interface NlqMessage {
  role: 'assistant';
  content: string; // stringified json
  type: 'nlq';
}
interface RegularMessage {
  role: 'user' | 'assistant';
  content: string;
}
export type ChatMessage = RegularMessage | NlqMessage | TextMessage;

export interface Chat {
  chatId: string;
  contextId: string;
  contextTitle: string;
  contextDetails?: ChatContextDetails;
  expireAt: string;
  tenantId: string;
  userId: string;
  chatHistory: ChatMessage[];
}
export type ChatWithoutHistory = Omit<Chat, 'chatHistory'>;

export interface LlmConfig {
  provider: string;
  apiKey: string;
  baseUrl: string;
  model: string;
  version?: string;
}

export interface ChatRequest {
  text?: string;
  timezone?: string;
  options?: {
    enableFollowup?: boolean;
  };
}

interface TextResponse {
  chatHistory: ChatMessage[];
  data: {
    answer: string;
  };
  chatId: string;
  responseType: 'text';
}

interface ErrorResponse {
  data: {
    message: string;
    code: string;
    answer: string;
  };
  responseType: 'error';
}

/**
 * Response data for NLQ request
 */
export interface NlqResponseData extends ExpandedQueryModel {
  /** Detailed description of the response */
  detailedDescription: string;
  /** An array of suggested followup questions  */
  followupQuestions?: string[];
  /** NLQ prompt used in the request  */
  nlqPrompt: string;
  /** Suggested widget props*/
  widgetProps?: WidgetProps | undefined;
  /** The response message for the chat  */
  userMsg: string;
  /** @internal */
  clarification?: string;
}
export interface NlqResponse {
  data: NlqResponseData;
  timestamp: string;
  responseType: 'nlq';
}
export type ChatResponse = NlqResponse | TextResponse | ErrorResponse;

export interface QueryRecommendationConfig {
  numOfRecommendations: number;
  userPrompt?: string;
}

export type NlqResult = Omit<NlqResponseData, 'followupQuestions'>;

/** AI-generated recommended query you can run on your data model */
export interface QueryRecommendation extends Omit<NlqResponseData, 'followupQuestions'> {}
export type QueryRecommendationResponse = QueryRecommendation[];

export interface GetNlgInsightsRequest {
  jaql: Pick<JaqlQueryPayload, 'datasource' | 'metadata' | 'filterRelations'>;
  verbosity?: 'Low' | 'High';
}

/** @internal */
export interface GetNlgInsightsResponse {
  data?: {
    answer: string;
  };
  responseType: 'Text';
}

/** @internal */
export const NLQ_RESULT_CHART_TYPES = [
  'indicator',
  'column',
  'line',
  'table',
  'pie',
  'bar',
] as const;
/** NLQ result chart types @expandType */
export type NlqResultChartType = (typeof NLQ_RESULT_CHART_TYPES)[number];
export interface GetNlqResultRequest {
  text: string;
  timezone?: string;
  chartTypes: NlqResultChartType[];
}

export interface SendFeedbackRequest {
  type: string;
  data: object;
  sourceId: string;
  rating: -1 | 1;
}

/**
 * The chat mode to use for a chat session
 *
 * analyze: Enable business users to uncover data insights
 * develop: Enable developers to build queries and charts for embedded analytics
 */
export type ChatMode = 'analyze' | 'develop';
