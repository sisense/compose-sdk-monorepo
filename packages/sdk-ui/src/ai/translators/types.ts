import { MetadataItem } from '@sisense/sdk-query-client';
import { AxesMapping, NlqResponseData } from '@/ai';
import { ChartType, TableType } from '@/types';
import { WidgetModel } from '@/models';

/**
 * Expanded Query Model that is based on NlqResponseData.
 * It contains expanded JAQL and chart recommendations returned from the chat response.
 *
 * @internal
 */
export type ExpandedQueryModel = Pick<
  NlqResponseData,
  'jaql' | 'chartRecommendations' | 'queryTitle'
>;

/**
 * Expanded Query Model
 *
 * Used as a default value for ExpandedQueryModel
 *
 * @internal
 */
export const EMPTY_EXPANDED_QUERY_MODEL: ExpandedQueryModel = {
  jaql: { datasource: { title: '' }, metadata: [] },
  queryTitle: '',
  chartRecommendations: {},
};

/**
 * Simple Chart Recommendations
 *
 * @internal
 */
export interface SimpleChartRecommendations {
  chartType: string;
  dataOptions: AxesMapping;
}

/**
 * Simple Query Model that is a simplified version of ExpandedQueryModel.
 *
 * It removes any JAQL syntax or information that already exists in the schema.
 * Also, any parameters that have a default do not need to be specified.
 *
 * Users manipulates this model via the query in YAML format.
 *
 * ModelTranslator is responsible for translating this query model to SimpleWidgetModel.
 *
 * @internal
 */
export interface SimpleQueryModel {
  /**
   * The data model or perspective title.
   */
  model: string;
  metadata: MetadataItem[];
  chart: SimpleChartRecommendations | {};
  /**
   * Title of the query
   */
  queryTitle: string;
}

/**
 * Empty Simple Query Model
 *
 * Used as a default value for SimpleQueryModel
 *
 * @internal
 */
export const EMPTY_SIMPLE_QUERY_MODEL: SimpleQueryModel = {
  model: '',
  metadata: [],
  chart: {},
  queryTitle: '',
};

/**
 * All Chart Types including Table
 * Note: Table will be re-united with ChartType in a future MR
 *
 * @internal
 */
export type AllChartType = ChartType | TableType;

/**
 * Chart Widget Model
 *
 * This model serves two purposes:
 *
 * (1) For rendering a chart or table widget in the Query Composer.
 * (2) For reverse engineered (translated) to CSDK code
 *
 * Naming TBD
 *
 * @internal
 */
export type ChartWidgetModel = Pick<WidgetModel, 'getChartWidgetProps' | 'getTableWidgetProps'> & {
  chartType: AllChartType;
  queryTitle: string;
};

/**
 * UI Framework
 *
 * @internal
 */
export type UiFramework = 'react' | 'vue' | 'angular';

/**
 * Code Template Key
 *
 * @internal
 */
export type CodeTemplateKey = 'baseChartTmpl' | 'chartTmpl' | 'chartWidgetTmpl' | 'tableWidgetTmpl';

/**
 * Code Templates
 *
 * @internal
 */
export type CodeTemplates = {
  [key in UiFramework]: { [key in CodeTemplateKey]: string };
};
