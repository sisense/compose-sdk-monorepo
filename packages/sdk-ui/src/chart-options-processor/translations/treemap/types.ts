import { Chart, PointLabelObject } from '@sisense/sisense-charts';

export type ParentValues = {
  direction: number;
  height: number;
  val: number;
  width: number;
  x: number;
  y: number;
};

export type LayoutPointResult = {
  x: number;
  y: number;
  width: number;
  height: number;
};

export type TreemapLayoutAlgorithmContext = {
  chart: Chart;
  squarified: (parent: ParentValues, children: PointLabelObject[]) => LayoutPointResult[];
};
