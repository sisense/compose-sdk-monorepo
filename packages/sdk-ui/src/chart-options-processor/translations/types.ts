import { AreamapChartDesignOptions } from '@/chart/restructured-charts/areamap-chart/types';
import { ChartType } from '../../types';
import {
  LineChartDesignOptions,
  AreaChartDesignOptions,
  BarChartDesignOptions,
  ColumnChartDesignOptions,
  CalendarHeatmapChartDesignOptions,
  PieChartDesignOptions,
  FunnelChartDesignOptions,
  IndicatorChartDesignOptions,
  PolarChartDesignOptions,
  ScatterChartDesignOptions,
  TreemapChartDesignOptions,
  BoxplotChartDesignOptions,
  ScattermapChartDesignOptions,
  AreaRangeChartDesignOptions,
  BaseDesignOptionsType,
} from './design-options';

/**
 * Style configuration for text elements in charts and UI components.
 * Defines the visual appearance and behavior of text rendering.
 */
export type TextStyle = {
  /**
   * Font family name for the text.
   * Examples: 'Arial', 'Helvetica', 'Times New Roman', 'sans-serif'
   */
  fontFamily?: string;
  /**
   * Font size in CSS units (px, em, rem, pt, etc.).
   * Examples: '12px', '1.2em', '16pt'
   */
  fontSize?: string;
  /**
   * Font weight for text rendering.
   * Examples: 'normal', 'bold', '100', '200', '300', '400', '500', '600', '700', '800', '900'
   */
  fontWeight?: string;
  /**
   * Font style for text rendering.
   * Examples: 'normal', 'italic', 'oblique'
   */
  fontStyle?: string;
  /**
   * Text color in any valid CSS color format.
   * Examples: '#FF0000', 'rgb(255, 0, 0)', 'red', 'rgba(255, 0, 0, 0.5)'
   */
  color?: string;
  /**
   * Text outline/stroke configuration.
   * Defines the outline color and width for text rendering.
   * Examples: '1px solid black', '2px #333'
   */
  textOutline?: string;
  /**
   * CSS pointer-events property for text interaction.
   * Controls how the text responds to mouse events.
   * Examples: 'auto', 'none', 'visible', 'visibleFill', 'visibleStroke'
   */
  pointerEvents?: string;
  /**
   * Text overflow handling behavior.
   * Controls how text is displayed when it exceeds its container.
   * Examples: 'clip', 'ellipsis', 'visible', 'hidden'
   */
  textOverflow?: string;
};

export const POLAR_CHART_TYPES = ['polar'] as const;
export type PolarChartType = (typeof POLAR_CHART_TYPES)[number];

export const CARTESIAN_CHART_TYPES = [
  'line',
  'area',
  'bar',
  'column',
  ...POLAR_CHART_TYPES,
] as const;
/** Cartesian family of chart types @expandType */
export type CartesianChartType = (typeof CARTESIAN_CHART_TYPES)[number];

export const CATEGORICAL_CHART_TYPES = ['pie', 'funnel', 'treemap', 'sunburst'] as const;
/** Categorical family of chart types  @expandType */
export type CategoricalChartType = (typeof CATEGORICAL_CHART_TYPES)[number];

export const SCATTER_CHART_TYPES = ['scatter'] as const;
/** Scatter chart types  @expandType */
export type ScatterChartType = (typeof SCATTER_CHART_TYPES)[number];

export const TABLE_TYPES = ['table'] as const;
/** Table chart types @expandType */
export type TableChartType = (typeof TABLE_TYPES)[number];

/** Table chart types @expandType */
export type TableType = TableChartType;

const INDICATOR_CHART_TYPES = ['indicator'] as const;
/** Indicator chart types @expandType */
export type IndicatorChartType = (typeof INDICATOR_CHART_TYPES)[number];

export const BOXPLOT_CHART_TYPES = ['boxplot'] as const;
/** Boxplot chart types  @expandType */
export type BoxplotChartType = (typeof BOXPLOT_CHART_TYPES)[number];

export const AREAMAP_CHART_TYPES = ['areamap'] as const;
/** Areamap chart types @expandType */
export type AreamapChartType = (typeof AREAMAP_CHART_TYPES)[number];

const IMAGE_CHART_TYPES = ['image'] as const;
export type ImageChartType = (typeof IMAGE_CHART_TYPES)[number];

export const SCATTERMAP_CHART_TYPES = ['scattermap'] as const;
/** Scattermap chart types  @expandType */
export type ScattermapChartType = (typeof SCATTERMAP_CHART_TYPES)[number];

export const CALENDAR_HEATMAP_CHART_TYPES = ['calendar-heatmap'] as const;
/**
 * Calendar heatmap chart types
 *
 * @expandType
 */
export type CalendarHeatmapChartType = (typeof CALENDAR_HEATMAP_CHART_TYPES)[number];

export const RANGE_CHART_TYPES = ['arearange'] as const;
/** AreaRange chart types  @expandType */
export type RangeChartType = (typeof RANGE_CHART_TYPES)[number];

// ChartDataType is the category of data structure for a group of charts,
// e.g. the ChartDataType of both "line" and "bar" charts is "cartesian",
// but a bubble/scatter chart would be a different data type.
export type ChartDataType =
  | 'cartesian'
  | 'categorical'
  | 'scatter'
  | 'table'
  | 'indicator'
  | 'areamap'
  | 'arearange';

/**
 * Design options for a specific chart type
 */
export type DesignOptions<SpecificChartType extends ChartType = ChartType> =
  SpecificChartType extends 'line'
    ? LineChartDesignOptions
    : SpecificChartType extends 'area'
    ? AreaChartDesignOptions
    : SpecificChartType extends 'bar'
    ? BarChartDesignOptions
    : SpecificChartType extends 'column'
    ? ColumnChartDesignOptions
    : SpecificChartType extends 'pie'
    ? PieChartDesignOptions
    : SpecificChartType extends 'funnel'
    ? FunnelChartDesignOptions
    : SpecificChartType extends 'polar'
    ? PolarChartDesignOptions
    : SpecificChartType extends 'indicator'
    ? IndicatorChartDesignOptions
    : SpecificChartType extends 'scatter'
    ? ScatterChartDesignOptions
    : SpecificChartType extends 'treemap'
    ? TreemapChartDesignOptions
    : SpecificChartType extends 'sunburst'
    ? TreemapChartDesignOptions
    : SpecificChartType extends 'boxplot'
    ? BoxplotChartDesignOptions
    : SpecificChartType extends 'areamap'
    ? AreamapChartDesignOptions
    : SpecificChartType extends 'scattermap'
    ? ScattermapChartDesignOptions
    : SpecificChartType extends 'calendar-heatmap'
    ? CalendarHeatmapChartDesignOptions
    : SpecificChartType extends 'arearange'
    ? AreaRangeChartDesignOptions
    : never;

/** A unique identifier for a series to be found in {@link ChartDataOptionsInternal} */
type SeriesId = string;

/**
 * Design options for a chart.
 * This includes global design options and specific design options per series.
 */
export type ChartDesignOptions<SpecificChartType extends ChartType = ChartType> =
  DesignOptions<SpecificChartType>;

/**
 * Design options, limited to the only series relevant options.
 *
 * @internal
 */
export type SeriesDesignOptions = Pick<BaseDesignOptionsType, 'marker'> &
  Pick<LineChartDesignOptions, 'line'>;
export type DesignPerSeries = { designPerSeries: Record<SeriesId, SeriesDesignOptions> };

export const isCartesian = (chartType: ChartType): chartType is CartesianChartType => {
  // Use .find instead of .includes for a more flexible type signature
  return CARTESIAN_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isCategorical = (chartType: ChartType): chartType is CategoricalChartType => {
  // Use .find instead of .includes for a more flexible type signature
  return CATEGORICAL_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isScatter = (chartType: ChartType): chartType is ScatterChartType => {
  // Use .find instead of .includes for a more flexible type signature
  return SCATTER_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isScattermap = (chartType: ChartType): chartType is ScattermapChartType => {
  // Use .find instead of .includes for a more flexible type signature
  return SCATTERMAP_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isIndicator = (chartType: ChartType): chartType is 'indicator' => {
  return INDICATOR_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isPolar = (chartType: ChartType): chartType is PolarChartType => {
  return POLAR_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isTable = (chartType: ChartType | TableType): chartType is TableType => {
  return TABLE_TYPES.find((value) => value === chartType) !== undefined;
};

export const isBoxplot = (chartType: ChartType): chartType is BoxplotChartType => {
  return BOXPLOT_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isAreamap = (chartType: ChartType): chartType is AreamapChartType => {
  return AREAMAP_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isRange = (chartType: ChartType): chartType is RangeChartType => {
  return RANGE_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isCalendarHeatmap = (chartType: ChartType): chartType is CalendarHeatmapChartType => {
  return CALENDAR_HEATMAP_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const ALL_CHART_TYPES = [
  ...CARTESIAN_CHART_TYPES,
  ...CATEGORICAL_CHART_TYPES,
  ...SCATTER_CHART_TYPES,
  ...TABLE_TYPES,
  ...INDICATOR_CHART_TYPES,
  ...BOXPLOT_CHART_TYPES,
  ...AREAMAP_CHART_TYPES,
  ...IMAGE_CHART_TYPES,
  ...SCATTERMAP_CHART_TYPES,
  ...CALENDAR_HEATMAP_CHART_TYPES,
  ...RANGE_CHART_TYPES,
] as const;

export type DynamicChartType = (typeof ALL_CHART_TYPES)[number];

/*
export const isImage = (chartType: ChartType): chartType is ImageChartType => {
  return IMAGE_CHART_TYPES.find((value) => value === chartType) !== undefined;
};
*/
