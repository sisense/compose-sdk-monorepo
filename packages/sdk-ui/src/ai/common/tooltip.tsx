import { Tooltip as MuiTooltip, type TooltipProps as MuiTooltipProps } from '@mui/material';

import { colors } from '@/themes/colors';

type TooltipProps = Pick<MuiTooltipProps, 'title' | 'children'>;

export default function Tooltip({ title, children }: TooltipProps) {
  return (
    <MuiTooltip
      title={title}
      placement="top"
      componentsProps={{
        tooltip: {
          sx: {
            bgcolor: colors.background.workspace,
            color: colors.text.content,
            paddingX: '16px',
            paddingY: '12px',
            fontSize: '13px',
            fontFamily: 'Open Sans',
            fontWeight: 400,
            borderRadius: '4px',
            boxShadow:
              '0px 4px 12px 0px rgba(9, 9, 10, 0.20), 0px 1px 4px 0px rgba(9, 9, 10, 0.10);',
          },
        },
        arrow: {
          sx: {
            color: colors.background.workspace,
          },
        },
      }}
      arrow
    >
      {children}
    </MuiTooltip>
  );
}
