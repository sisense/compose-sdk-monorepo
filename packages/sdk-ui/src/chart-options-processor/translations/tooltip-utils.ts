import { isNumber } from '@sisense/sdk-data';
import { Category, Value, isValue } from '../../chart-data-options/types.js';
import { applyFormat, defaultConfig } from './number-format-config.js';

export const isXValueNumeric = (dataOptionX: Value | Category | undefined) =>
  dataOptionX ? isValue(dataOptionX) || (dataOptionX.type && isNumber(dataOptionX.type)) : false;

export const formatTooltipValue = (
  dataOption: Value | Category | undefined,
  value: number | undefined,
  displayValue: string,
) => {
  if (!dataOption || value === undefined || isNaN(value)) return displayValue;

  return dataOption.numberFormatConfig
    ? applyFormat(dataOption.numberFormatConfig, value)
    : applyFormat(defaultConfig, value);
};

export const formatTooltipXValue = (
  dataOption: Value | Category | undefined,
  value: number | string | undefined,
  displayValue: string,
) => {
  const isNumeric = isXValueNumeric(dataOption);

  return isNumeric
    ? formatTooltipValue(dataOption, parseFloat(`${value}`), displayValue)
    : displayValue;
};

export type TooltipSettings = {
  enabled?: boolean;
  animation?: boolean;
  backgroundColor?: string;
  borderColor?: string;
  borderRadius?: number;
  borderWidth?: number;
  useHTML?: boolean;
  formatter?: () => string;
  style?: {
    fontFamily?: string;
  };
};

export type InternalSeries = {
  series: { name: string; color: string };
  x: string;
  y: number;
  point: {
    x?: number | string;
    y?: number;
    z?: number;
    name: string;
    color: string;
    custom?: {
      number1?: number;
      string1?: string;
      xDisplayValue?: string;
      xValue?: (number | string)[];
    };
  };
  percentage?: number;
};
