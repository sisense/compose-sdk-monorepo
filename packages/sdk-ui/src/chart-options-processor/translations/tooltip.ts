/* eslint-disable max-lines-per-function */
/* eslint-disable complexity */
import {
  CartesianChartDataOptionsInternal,
  ChartDataOptionsInternal,
} from '../../chart-data-options/types';
import { colorChineseSilver, colorWhite } from '../chart-options-service';
import {
  InternalSeries,
  TooltipSettings,
  formatTooltipValue,
  formatTooltipXValue,
} from './tooltip-utils';

export const getTooltipSettings = (
  showDecimals: boolean | undefined,
  chartDataOptions: ChartDataOptionsInternal,
): TooltipSettings => {
  return {
    animation: false,
    backgroundColor: colorWhite,
    borderColor: colorChineseSilver,
    borderRadius: 10,
    borderWidth: 1,
    useHTML: true,
    formatter: function () {
      const that: InternalSeries = this as InternalSeries;

      // Applicable only to pie and funnel charts
      let percentage: string | undefined;
      if (that.percentage) {
        percentage = showDecimals ? that.percentage.toFixed(1) : `${Math.round(that.percentage)}`;
      }

      const cartesianChartDataOptions: CartesianChartDataOptionsInternal =
        chartDataOptions as CartesianChartDataOptionsInternal;

      const dataOptionY =
        cartesianChartDataOptions.breakBy.length > 0
          ? cartesianChartDataOptions.y?.find((y) => y.enabled)
          : cartesianChartDataOptions.y?.find((y) => y.title === that.series.name);

      const yValue = formatTooltipValue(dataOptionY, that.point.y, '');

      const maskedX = that.point?.custom?.xDisplayValue ?? that.x;
      const x1Value = cartesianChartDataOptions.x
        ? formatTooltipXValue(cartesianChartDataOptions.x[0], that.x, maskedX)
        : maskedX;

      let x2Value = undefined;
      if (
        that.point.custom?.xValue &&
        cartesianChartDataOptions?.x &&
        cartesianChartDataOptions.x.length === 2
      ) {
        const maskedX1 = `${that.point.custom.xValue[0]}`; // X2 is in position xValue[0]
        x2Value = formatTooltipXValue(cartesianChartDataOptions.x[1], maskedX1, maskedX1);
      }

      return `<div
                   style="minWidth: 100px; color: #acacac; fontSize: 13px; lineHeight: 20px">
              ${that.point.name || that.series.name}
              <br />
              <span style="color: ${that.point.color || that.series.color}">
              ${yValue}
              ${percentage ? ` / ${percentage}%` : ''}
            </span>
            ${x2Value ? `<hr style="margin: 5px 0px" />${x2Value}` : ''}
            ${x1Value ? `<hr style="margin: 5px 0px" />${x1Value}` : ''}
            </div>`;
    },
  };
};
