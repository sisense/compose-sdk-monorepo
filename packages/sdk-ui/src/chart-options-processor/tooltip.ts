/* eslint-disable complexity */
/* eslint-disable sonarjs/no-nested-template-literals */
import {
  CartesianChartDataOptionsInternal,
  ChartDataOptionsInternal,
} from '../chart-data-options/types';
import { colorChineseSilver, colorWhite } from './chart-options-service';
import { applyFormat, defaultConfig } from './translations/number-format-config';
import { TooltipOptions } from '@sisense/sisense-charts';
import { spanSegment, tooltipSeparator, tooltipWrapper } from './translations/scatter-tooltip';

export type TooltipSettings = TooltipOptions;

export type InternalSeriesNode = {
  val: number;
  name: string;
  parentNode?: InternalSeriesNode;
  color?: string;
};

export type InternalSeries = {
  series: { name: string; color: string };
  x: string;
  y: number;
  point: {
    name: string;
    color: string;
    custom?: { number1?: number; string1?: string; xDisplayValue?: string };
    node?: InternalSeriesNode;
  };
  percentage?: number;
  color?: string;
};

export const getTooltipSettings = (
  showDecimals: boolean | undefined,
  chartDataOptions: ChartDataOptionsInternal,
): TooltipSettings => {
  return {
    animation: false,
    backgroundColor: colorWhite,
    borderColor: colorChineseSilver,
    borderRadius: 10,
    borderWidth: 1,
    useHTML: true,
    formatter: function () {
      const that: InternalSeries = this as InternalSeries;

      // Applicable only to pie and funnel charts
      let percentage: string | undefined;
      if (that.percentage) {
        percentage = showDecimals ? that.percentage.toFixed(1) : `${Math.round(that.percentage)}`;
      }

      const cartesianChartDataOptions: CartesianChartDataOptionsInternal =
        chartDataOptions as CartesianChartDataOptionsInternal;
      let numberFormatConfig = cartesianChartDataOptions.y?.find(
        (y) => y.name === that.series.name,
      )?.numberFormatConfig;
      if (!numberFormatConfig && cartesianChartDataOptions.breakBy.length > 0) {
        numberFormatConfig = cartesianChartDataOptions.y?.find(
          (y) => y.enabled,
        )?.numberFormatConfig;
      }
      if (!numberFormatConfig) {
        numberFormatConfig = defaultConfig;
      }
      const xValue = that.point?.custom?.xDisplayValue ?? that.x;
      const value =
        applyFormat(numberFormatConfig, that.y) + (percentage ? ` / ${percentage}%` : '');
      const color = that.point.color || that.series.color;

      return tooltipWrapper(`
          ${that.series.name || ''}
          ${that.series.name && that.point.name ? ' - ' : ''}
          ${that.point.name || ''}
          <br />
          ${spanSegment(value, color)}
          ${xValue ? tooltipSeparator() + xValue : ''}
        `);
    },
  };
};
