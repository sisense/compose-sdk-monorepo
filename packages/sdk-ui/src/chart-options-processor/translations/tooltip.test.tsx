import type { NumberFormatConfig } from '@/types';
import { CartesianChartDataOptionsInternal } from '../../chart-data-options/types';
import { getTooltipSettings } from './tooltip';
import { InternalSeries } from './tooltip-utils';

const format1: NumberFormatConfig = {
  name: 'Currency',
  symbol: '$',
  decimalScale: 2,
};
const format2: NumberFormatConfig = {
  name: 'Currency',
  symbol: '!',
  decimalScale: 3,
};
const format3: NumberFormatConfig = {
  name: 'Currency',
  symbol: '@',
  decimalScale: 4,
};

const dataOptions: CartesianChartDataOptionsInternal = {
  x: [
    { name: 'x1', type: 'number', numberFormatConfig: format1 },
    { name: 'x2', type: 'number', numberFormatConfig: format2 },
  ],
  y: [{ title: 'v', name: 'v', numberFormatConfig: format3, enabled: true }],
  breakBy: [{ name: 'b', type: 'number' }],
};

it('should display cartesian tooltip for point', () => {
  const point: InternalSeries = {
    series: { name: '3.14', color: 'red' },
    x: '1.25905',
    y: 42.0009,
    point: {
      x: 1,
      y: 42.0009,
      name: '3.14',
      color: 'red',
      custom: { xValue: [9.8765, 1.25905] },
    },
  };

  const seriesPoint = { ...point, ...getTooltipSettings(false, dataOptions) };
  const tooltip = seriesPoint.formatter ? seriesPoint.formatter() : null;
  expect(tooltip).toMatchSnapshot();
});

it('should display pie tooltip for point', () => {
  const point: InternalSeries = {
    series: { name: '3.14', color: 'red' },
    x: '',
    y: 42.0009,
    percentage: 20,
    point: {
      y: 42.0009,
      name: '3.14',
      color: 'red',
      custom: {},
    },
  };

  const seriesPoint = {
    ...point,
    ...getTooltipSettings(false, { ...dataOptions, breakBy: [dataOptions.x[0]] }),
  };
  const tooltip = seriesPoint.formatter ? seriesPoint.formatter() : null;
  expect(tooltip).toMatchSnapshot();
});
