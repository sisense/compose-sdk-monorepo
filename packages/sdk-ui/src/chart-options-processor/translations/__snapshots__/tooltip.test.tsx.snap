// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should display cartesian tooltip for point 1`] = `
"<div
                   style=\\"minWidth: 100px; color: #acacac; fontSize: 13px; lineHeight: 20px\\">
              3.14
              <br />
              <span style=\\"color: red\\">
              @42.0009

            </span>
            <hr style=\\"margin: 5px 0px\\" />!9.877
            <hr style=\\"margin: 5px 0px\\" />$1.26
            </div>"
`;

exports[`should display pie tooltip for point 1`] = `
"<div
                   style=\\"minWidth: 100px; color: #acacac; fontSize: 13px; lineHeight: 20px\\">
              3.14
              <br />
              <span style=\\"color: red\\">
              @42.0009
               / 20%
            </span>


            </div>"
`;
